File: solutions/1/java/solution.java
import java.util.*;

class Solution {
    public int[] twoNum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();

        for (int i = 0; i < nums.length; i++) {
            int requiredNum = target - nums[i];
            if (map.containsKey(requiredNum)) {
                return new int[] { map.get(requiredNum), i };
            }
            map.put(nums[i], i);
        }

        return new int[] {};
    }
}

File: solutions/1/javascript/solution.js
// Should return indices of the two numbers such that they add up to target.

// Old version - the problem is that, this solution has a complexity of O(n^2)
/* const twoSum = (nums, target) => {
	for (let i = 0; i < nums.length; i++) {
		let aux = target - nums[i];
		for (let j = i + 1; j < nums.length; j++) {
			if (nums[j] === aux) return [i, j];
		}
	}
}; */

// Better version - with a complexity of O(n)
const twoSum = (nums, target) => {
	const map = new Map();

	for (let i = 0; i < nums.length; i++) {
		let requiredNum = target - nums[i];
		if (map.has(requiredNum)) return [map.get(requiredNum), i];
		map.set(nums[i], i);
	}
};
